libyul/YulName.h

```diff
@@ -18,9 +18,114 @@
+ *
+ *  - 'defined', i.e., they are equipped with a string label, which can be retrieved by a call to `labelOf(yulName)`;
+ *  - 'derived', i.e., they don't possess a string label but have a parent yul name which is either also derived or
+ *    defined. All dependency chains of derived labels terminate in a defined label.
+ *
+ * Such derived identifiers can be introduced, e.g., during certain optimization steps. If the AST (or segments thereof)
+ * should be printed and/or the string label of identifiers should be retrieved which are still in derived state,
+ * a call to `generateLabels(identifiers)` changes the status of all derived labels to defined and generates
+ * unique labels for all identifiers provided to the method. The generated labels are based on their parents.
+ */
+class YulNameRepository
+{
+public:
+	using YulName = std::uint64_t;
+	YulNameRepository();
+	/// Defines a new name based on a label. If the label already was defined, it returns the corresponding YulName
+	/// instead of a new one (`defineName("xyz") == defineName("xyz")`).
+	YulName defineName(std::string_view _label)
+	{
+		if (auto const name = nameOfLabel(_label); name.has_value())
+			return *name;
+
+		m_definedLabels.emplace_back(_label);
+		m_names.emplace_back(m_definedLabels.size() - 1, YulNameState::DEFINED);
+		return static_cast<YulName>(m_names.size() - 1);
+	}
+
+	/// Defines a new name based on a parent name. When generating labels, the generated label will be based on the
+	/// parent's (`deriveName(id) != deriveName(id)`).
+	YulName deriveName(YulName const _baseName)
+	{
+		m_names.emplace_back(baseNameOf(_baseName), YulNameState::DERIVED);
+		return static_cast<YulName>(m_names.size() - 1);
+	}
+
+	/// The empty name.
+	static constexpr YulName emptyName() { return 0; }
+
+	/// Yields the label of a yul name. The name must have been added via `defineName`, a label must have been
+	/// generated with `generateLabels`, or it is a builtin.
+	std::optional<std::string_view> labelOf(YulName const _name) const
+	{
+		if (isDerivedName(_name))
+			return std::nullopt;
+
+		auto const labelIndex = std::get<0>(m_names.at(static_cast<size_t>(_name)));
+		return m_definedLabels[static_cast<size_t>(labelIndex)];
+	}
+
+	/// If it can be assumed that the label was already generated, this function will yield it (or fail with an
+	/// assertion error).
+	std::string_view requiredLabelOf(YulName _name) const;
+
+	/// Yields the label of the base name of the provided name. Opposed to `labelOf`, this must always exist.
+	std::string_view baseLabelOf(YulName const _name) const
+	{
+		auto const labelIndex = std::get<0>(m_names.at(static_cast<size_t>(baseNameOf(_name))));
+		return m_definedLabels[static_cast<size_t>(labelIndex)];
+	}
+
+	/// Whether a name is considered derived, i.e., has no label but a parent name.
+	bool isDerivedName(YulName const _name) const
+	{
+		return std::get<1>(m_names.at(static_cast<size_t>(_name))) == YulNameState::DERIVED;
```

Comment:
It is smaller, hence all the static casts. Perhaps we should restrict the id type to 32 bit to begin with.

Reply:
If you're assuming that this can never exceed 32 bits then yeah, using `uint64_t` is pointless and misleading. I guess 4 billion names should be more than enough in practical use.

We could also use `size_t` if you think this value will never be exposed to the user and will not need to have its range validated. The point of the original suggestion to use a type with a platform-independent range was that we may need to expose it somewhere though.